### PR TITLE
feat(profiling): sync flamechart state with qs and intialize from it

### PR DESCRIPTION
**Pull request by JonasBa**
This PR adds syncing of flamegraph state with query string (and initializing from it), meaning flamegraph views can now be shared via URL. I lifted the importProfile code and activeProfileIndex out of the flamegraph to avoid reading it from multiple places and somewhat centralize it... In general, I'm not very happy with how state management is evolving (the side effects are getting hard to track) and not being able to test this is starting to weigh down on us (webgl initialize calls fail in node). 

I dont expect us to do many more changes to the state management in the near future, but we should think about how we can improve it.

CleanShot 2022-04-26 at 14 21 32

### CODE REVIEW

File: static/app/components/profiling/flamegraph.tsx [hunk 1]
```diff
@@ -24,44 +28,38 @@ function getTransactionConfigSpace(profiles: Profile[]): Rect {
...
 
   const flamegraph = useMemo(() => {
-    // if the activeProfileIndex is null, use the activeProfileIndex from the profile group
-    const profileIndex = activeProfileIndex ?? profiles.activeProfileIndex;
+    if (
+      typeof activeProfileIndex !== 'number' ||
+      props.profiles.profiles[activeProfileIndex] === undefined
+    ) {
+      return FlamegraphModel.Empty();
+    }
```
Comment (Zylphrex) [inline static/app/components/profiling/flamegraph.tsx:48]: I don't think you need an empty flame graph here. When the active profile index from the state is `null`, you should be using `props.profiles.activeProfileIndex`. This should also simplify a lot of the code below where you handle this empty state.

File: static/app/components/profiling/flamegraph.tsx [hunk 2]
```diff
@@ -78,8 +76,10 @@ function Flamegraph(props: FlamegraphProps): ReactElement {
         />
         <ThreadMenuSelector
           profileGroup={props.profiles}
-          activeProfileIndex={flamegraph.profileIndex}
-          onProfileIndexChange={setActiveProfileIndex}
+          activeProfileIndex={activeProfileIndex ?? 0}
```
Comment (Zylphrex) [inline static/app/components/profiling/flamegraph.tsx:79]: If you make the above change, you should be able to go back to using `flamegraph.profileIndex`

File: static/app/components/profiling/flamegraphZoomView.tsx [hunk 3]
```diff
@@ -68,6 +68,18 @@ function FlamegraphZoomView({
           {draw_border: true}
         );
 
+        // If we have previous position on the flamechart, apply it
```
Comment (Zylphrex) [inline static/app/components/profiling/flamegraphZoomView.tsx:71]: Is this to populate the position from the URL?

File: static/app/components/profiling/flamegraph.tsx [hunk 4]
```diff
@@ -91,9 +91,8 @@ function Flamegraph(props: FlamegraphProps): ReactElement {
         />
       </FlamegraphZoomViewMinimapContainer>
       <FlamegraphZoomViewContainer>
-        <ProfileDragDropImport onImport={onImport}>
+        <ProfileDragDropImport onImport={props.onImport}>
           <FlamegraphZoomView
-            key={`${profiles.traceID}-${flamegraph.profileIndex}`}
```
Comment (Zylphrex) [inline static/app/components/profiling/flamegraph.tsx:96]: Why was this removed?

Comment (JonasBa) [reply static/app/components/profiling/flamegraph.tsx:96]: It was more of a "hack" before and it's annoying because it makes the underlying hooks inside the view behave slightly differently (like useEffect fires on mount each time they key changes instead of just prop changes etc), so it was adding to the confusion

File: static/app/components/profiling/flamegraph.tsx (hunk 1 above)
Comment (JonasBa) [reply static/app/components/profiling/flamegraph.tsx:48]: I dont think this makes any difference to the end user tbh. What I wanted to avoid was reading from two different sources to avoid confusion. The correct way to handle this would be to have a "thread not found" screen instead of the empty view, but I want to tackle that separately

File: static/app/components/profiling/flamegraph.tsx (hunk 4 above)
Comment (JonasBa) [reply static/app/components/profiling/flamegraph.tsx:96]: Ah, let me check that.

Comment (JonasBa) [reply static/app/components/profiling/flamegraph.tsx:100]: Manually reset this now when flamegraph renderer changes, but I will centralize all of this logic in a separate PR, these side effects and all are becoming painful to manage and track

File: static/app/utils/profiling/flamegraph/flamegraphStateProvider/index.tsx [hunk 5]
```diff
@@ -1,31 +1,138 @@
...
+function isColorCoding(
+  value: PossibleQuery['colorCoding'] | FlamegraphState['preferences']['colorCoding']
+): value is FlamegraphState['preferences']['colorCoding'] {
+  return (
+    value === 'by symbol name' ||
+    value === 'by library' ||
+    value === 'by recursion' ||
+    value === 'by system / application'
+  );
+}
```
Comment (Zylphrex) [inline static/app/utils/profiling/flamegraph/flamegraphStateProvider/index.tsx:37]: All these functions are kind of annoying if we add new options later right? Since typescript won't warn us that these definitions are no longer correct if a new option is added.

File: static/app/utils/profiling/renderers/flamegraphRenderer.tsx [hunk 6]
```diff
@@ -577,6 +577,8 @@ class FlamegraphRenderer {
     const physicalToConfig = mat3.invert(mat3.create(), configViewToPhysicalSpace);
     const configSpacePixel = physicalSpacePixel.transformRect(physicalToConfig);
 
+    // Reset the uniform values from previous draw
+    this.gl.uniform1i(this.uniforms.u_is_search_result, 0);
```
Comment (Zylphrex) [inline static/app/utils/profiling/renderers/flamegraphRenderer.tsx:581]: Dont need this anymore

File: static/app/views/profiling/flamegraph.tsx [hunk 7]
```diff
@@ -71,11 +79,37 @@ interface FlamegraphViewProps {
...
+          0,
+      },
+    };
+  }, [requestState, location.query]);
+
+  // Sometimes the page we are coming from had scrollTop > 0, so we need to
+  // force restore it to top of page because scrolling on a flamegraph does not scroll the page
+  useEffect(() => {
+    document.scrollingElement?.scrollTo(0, 0);
+  }, []);
```
Comment (Zylphrex) [inline static/app/views/profiling/flamegraph.tsx:111]: Do we still need this? The page no longer scrolls at all now that the footer is gone.

File: static/app/utils/profiling/flamegraph/flamegraphStateProvider/index.tsx [hunk 8]
```diff
@@ -1,31 +1,138 @@
...
-type FlamegraphState = React.ReducerState<FlamegraphStateReducer>['current'];
-
+export type FlamegraphState = React.ReducerState<FlamegraphStateReducer>['current'];
 type FlamegraphAction = React.Dispatch<
   UndoableReducerAction<React.ReducerAction<FlamegraphStateReducer>>
 >;
 
 type FlamegraphStateReducer = UndoableReducer<typeof combinedReducers>;
 
+export function FlamegraphStateQueryParamSync() {
```
Comment (Zylphrex) [inline static/app/utils/profiling/flamegraph/flamegraphStateProvider/index.tsx:116]: Should this be a hook instead a component?
